File: rust/src/http_client/advanced_client.rs
// #exonware/xwsystem/rust/src/http_client/advanced_client.rs
//! Company: eXonware.com
//! Author: eXonware Backend Team
//! Email: connect@exonware.com
//! Version: 0.1.0.1
//! Generation Date: September 04, 2025
//! 
//! Advanced HTTP client with HTTP/2, streaming, pluggable transports, and modern features.


use std::collections::HashMap;

use crate::http_client::client::{HttpError, RetryConfig};
use crate::config::logging_setup::get_logger;
use crate::http_client::contracts::ITransport as Transport;

/// Configuration for streaming operations.
#[derive(Debug, Clone)]
pub struct StreamingConfig {
    pub chunk_size: i64,
    pub buffer_size: i64,
    pub timeout_per_chunk: f64,
    pub max_content_length: Option<i64>,
}

impl Default for StreamingConfig {
    fn default() -> Self {
        Self {
            chunk_size: 8192,
            buffer_size: 65536,
            timeout_per_chunk: 30.0,
            max_content_length: None,
        }
    }
}

impl StreamingConfig {
    pub fn new() -> Self {
        Self::default()
    }
}

/// Configuration for HTTP/2 features.
#[derive(Debug, Clone)]
pub struct Http2Config {
    pub enabled: bool,
    pub max_concurrent_streams: i64,
    pub initial_window_size: i64,
    pub max_frame_size: i64,
    pub enable_push: bool,
}

impl Default for Http2Config {
    fn default() -> Self {
        Self {
            enabled: true,
            max_concurrent_streams: 100,
            initial_window_size: 65536,
            max_frame_size: 16384,
            enable_push: false,
        }
    }
}

impl Http2Config {
    pub fn new() -> Self {
        Self::default()
    }
}

/// Advanced HTTP client configuration.
#[derive(Debug, Clone)]
pub struct AdvancedHttpConfig {
    pub http2: Http2Config,
    pub streaming: StreamingConfig,
    pub retry: crate::http_client::client::RetryConfig,
    pub verify_ssl: bool,
    pub trust_env: bool,
    pub follow_redirects: bool,
    pub max_redirects: i64,
}

impl Default for AdvancedHttpConfig {
    fn default() -> Self {
        Self {
            http2: Http2Config::default(),
            streaming: StreamingConfig::default(),
            retry: crate::http_client::client::RetryConfig::default(),
            verify_ssl: true,
            trust_env: true,
            follow_redirects: true,
            max_redirects: 20,
        }
    }
}

impl AdvancedHttpConfig {
    pub fn new() -> Self {
        Self::default()
    }
}

// For simplicity, delegate to async version
/// Mock transport for testing purposes.
pub struct MockTransport {
    pub responses: HashMap<String, serde_json::Value>,
}

impl MockTransport {
    /// Initialize mock transport.
    ///
    ///
    /// Args:
    /// responses: Dictionary mapping URLs to mock responses
    pub fn new(
        responses: HashMap<String, serde_json::Value>
    ) -> Self {
        Self {
            responses,
        }
    }

    /// Handle async request with mock response.
    pub async fn handle_async_request(&self, request: serde_json::Value) -> MockResponse {
        // Extract URL from request (simplified - in production would parse properly)
        let url = request.get("url")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();
        
        if let Some(response_data) = self.responses.get(&url) {
            let status_code = response_data.get("status_code")
                .and_then(|v| v.as_i64())
                .unwrap_or(200);
            let content = response_data.get("content")
                .and_then(|v| v.as_str())
                .map(|s| s.as_bytes().to_vec())
                .unwrap_or_default();
            let headers = response_data.get("headers")
                .and_then(|v| v.as_object())
                .map(|obj| {
                    obj.iter()
                        .filter_map(|(k, v)| v.as_str().map(|s| (k.clone(), s.to_string())))
                        .collect()
                })
                .unwrap_or_default();
            
            return MockResponse::new(status_code, content, Some(headers), Some(url));
        }
        
        // Default 404 response
        MockResponse::new(404, b"Not Found".to_vec(), None, Some(url))
    }

    // For simplicity, delegate to async version
    /// Handle sync request with mock response.
    pub fn handle_request(&self, request: serde_json::Value) -> MockResponse {
        // In production, you'd use a runtime like tokio
        // For now, return a simplified sync version
        let url = request.get("url")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();
        
        if let Some(response_data) = self.responses.get(&url) {
            let status_code = response_data.get("status_code")
                .and_then(|v| v.as_i64())
                .unwrap_or(200);
            let content = response_data.get("content")
                .and_then(|v| v.as_str())
                .map(|s| s.as_bytes().to_vec())
                .unwrap_or_default();
            let headers = response_data.get("headers")
                .and_then(|v| v.as_object())
                .map(|obj| {
                    obj.iter()
                        .filter_map(|(k, v)| v.as_str().map(|s| (k.clone(), s.to_string())))
                        .collect()
                })
                .unwrap_or_default();
            
            return MockResponse::new(status_code, content, Some(headers), Some(url));
        }
        
        MockResponse::new(404, b"Not Found".to_vec(), None, Some(url))
    }

}

/// Mock HTTP response for testing.
pub struct MockResponse {
    pub status_code: i64,
    pub content: Vec<u8>,
    pub headers: HashMap<String, String>,
    pub url: String,
}

impl MockResponse {
    /// Constructor
    pub fn new(
        status_code: i64,
        content: Vec<u8>,
        headers: Option<HashMap<String, String>>,
        url: Option<String>
    ) -> Self {
        Self {
            status_code,
            content,
            headers,
            url,
        }
    }

    /// Parse JSON response.
    pub fn json(&self) -> Result<serde_json::Value, serde_json::Error> {
        let text = String::from_utf8_lossy(&self.content);
        serde_json::from_str(&text)
    }

    /// Raise exception for HTTP error status.
    pub fn raise_for_status(&self) -> Result<(), HttpError> {
        if 400 <= self.status_code && self.status_code < 600 {
            let message = format!("HTTP {}", self.status_code);
            let text = String::from_utf8_lossy(&self.content).to_string();
            return Err(HttpError::new(message));
        }
        Ok(())
    }

}

// Lazy installation system will handle httpx if missing
// Configure HTTP/2 if enabled
// Disable HTTP/2 with custom transports
// Configure SSL context
// Prepare request arguments
// Handle custom transport
// Prepare request arguments
// Custom transport doesn't support streaming
// Sync methods for compatibility
/// Advanced HTTP client with HTTP/2, streaming, pluggable transports, and modern features.
///
/// Features:
/// - HTTP/2 support with multiplexing
/// - Streaming request/response bodies
/// - Pluggable transport layer for testing
/// - Advanced connection management
/// - Comprehensive retry and error handling
/// - SSL/TLS configuration
/// - Request/response hooks
pub struct AdvancedHttpClient {
    pub base_url: Option<String>,
    pub timeout: f64,
    pub config: Option<AdvancedHttpConfig>,
    pub transport: Option<Transport>,
    pub default_headers: Option<HashMap<String, String>>,
}

impl AdvancedHttpClient {
    /// Initialize advanced HTTP client.
    ///
    ///
    /// Args:
    /// base_url: Base URL for all requests
    /// timeout: Request timeout in seconds
    /// config: Advanced configuration options
    /// transport: Custom transport (for testing/mocking)
    /// default_headers: Default headers for all requests
    pub fn new(
        base_url: Option<String>,
        timeout: Option<f64>,
        config: Option<AdvancedHttpConfig>,
        transport: Option<Transport>,
        default_headers: Option<HashMap<String, String>>
    ) -> Self {
        Self {
            base_url,
            timeout,
            config,
            transport,
            default_headers,
        }
    }

    /// Close sync client.
    pub fn close(&self) {
        // In production, you'd close the underlying HTTP client
        // For now, this is a placeholder
    }

    /// Close async client.
    pub async fn aclose(&self) {
        // In production, you'd close the underlying async HTTP client
        // For now, this is a placeholder
    }

    /// Async GET request.
    pub async fn get(&self, url: String, params: Option<HashMap<String, serde_json::Value>>, headers: Option<HashMap<String, String>>, kwargs: HashMap<String, String>) -> String
    {
        // TODO: This function uses external dependencies.
        // The original Python implementation should be maintained/adapted.
        //
        // Python → Rust crate mappings:
        //   httpx → reqwest
        //
        // Add these crates to Cargo.toml if implementing:
        //   reqwest = "*"
        todo!()
    }

    /// Async POST request.
    pub async fn post(&self, url: String, json: Option<serde_json::Value>, data: Option<serde_json::Value>, files: Option<HashMap<String, serde_json::Value>>, headers: Option<HashMap<String, String>>, kwargs: HashMap<String, String>) -> String
    {
        // TODO: This function uses external dependencies.
        // The original Python implementation should be maintained/adapted.
        //
        // Python → Rust crate mappings:
        //   httpx → reqwest
        //
        // Add these crates to Cargo.toml if implementing:
        //   reqwest = "*"
        todo!()
    }

    /// Async PUT request.
    pub async fn put(&self, url: String, json: Option<serde_json::Value>, data: Option<serde_json::Value>, headers: Option<HashMap<String, String>>, kwargs: HashMap<String, String>) -> String
    {
        // TODO: This function uses external dependencies.
        // The original Python implementation should be maintained/adapted.
        //
        // Python → Rust crate mappings:
        //   httpx → reqwest
        //
        // Add these crates to Cargo.toml if implementing:
        //   reqwest = "*"
        todo!()
    }

    /// Async PATCH request.
    pub async fn patch(&self, url: String, json: Option<serde_json::Value>, data: Option<serde_json::Value>, headers: Option<HashMap<String, String>>, kwargs: HashMap<String, String>) -> String
    {
        // TODO: This function uses external dependencies.
        // The original Python implementation should be maintained/adapted.
        //
        // Python → Rust crate mappings:
        //   httpx → reqwest
        //
        // Add these crates to Cargo.toml if implementing:
        //   reqwest = "*"
        todo!()
    }

    /// Async DELETE request.
    pub async fn delete(&self, url: String, headers: Option<HashMap<String, String>>, kwargs: HashMap<String, String>) -> String
    {
        // TODO: This function uses external dependencies.
        // The original Python implementation should be maintained/adapted.
        //
        // Python → Rust crate mappings:
        //   httpx → reqwest
        //
        // Add these crates to Cargo.toml if implementing:
        //   reqwest = "*"
        todo!()
    }

    /// Async HEAD request.
    pub async fn head(&self, url: String, headers: Option<HashMap<String, String>>, kwargs: HashMap<String, String>) -> String
    {
        // TODO: This function uses external dependencies.
        // The original Python implementation should be maintained/adapted.
        //
        // Python → Rust crate mappings:
        //   httpx → reqwest
        //
        // Add these crates to Cargo.toml if implementing:
        //   reqwest = "*"
        todo!()
    }

    /// Async OPTIONS request.
    pub async fn options(&self, url: String, headers: Option<HashMap<String, String>>, kwargs: HashMap<String, String>) -> String
    {
        // TODO: This function uses external dependencies.
        // The original Python implementation should be maintained/adapted.
        //
        // Python → Rust crate mappings:
        //   httpx → reqwest
        //
        // Add these crates to Cargo.toml if implementing:
        //   reqwest = "*"
        todo!()
    }

    // Prepare request arguments
    // Handle custom transport
    /// Make async HTTP request with retry logic.
    pub async fn request(&self, method: String, url: String, headers: Option<HashMap<String, String>>, params: Option<HashMap<String, serde_json::Value>>, json: Option<serde_json::Value>, data: Option<serde_json::Value>, files: Option<HashMap<String, serde_json::Value>>, kwargs: HashMap<String, String>) -> String
    {
        // TODO: This function uses external dependencies.
        // The original Python implementation should be maintained/adapted.
        //
        // Python → Rust crate mappings:
        //   client → (no known Rust equivalent)
        //   httpx → reqwest
        //   monitoring → (no known Rust equivalent)
        //   monitoring.error_recovery → (no known Rust equivalent)
        //
        // Add these crates to Cargo.toml if implementing:
        //   reqwest = "*"
        todo!()
    }

    // Prepare request arguments
    // Custom transport doesn't support streaming
    /// Stream HTTP request/response.
    ///
    ///
    /// Args:
    /// method: HTTP method
    /// url: Request URL
    /// headers: Request headers
    /// params: Query parameters
    /// json: JSON data
    /// data: Request body data
    /// **kwargs: Additional request arguments
    ///
    /// Yields:
    /// Streaming response object
    pub async fn stream(&self, method: String, url: String, headers: Option<HashMap<String, String>>, params: Option<HashMap<String, serde_json::Value>>, json: Option<serde_json::Value>, data: Option<serde_json::Value>, kwargs: HashMap<String, String>) -> ()
    {
        // TODO: Implement
        todo!()
    }

    /// Stream response content in chunks.
    ///
    ///
    /// Args:
    /// method: HTTP method
    /// url: Request URL
    /// chunk_size: Size of each chunk
    /// **kwargs: Additional request arguments
    ///
    /// Yields:
    /// Content chunks as bytes
    pub async fn stream_content(&self, method: String, url: String, chunk_size: Option<i64>, kwargs: HashMap<String, String>) -> String
    {
        // TODO: This function uses external dependencies.
        // The original Python implementation should be maintained/adapted.
        //
        // Python → Rust crate mappings:
        //   client → (no known Rust equivalent)
        //
        // Add these crates to Cargo.toml if implementing:
        todo!()
    }

    /// Stream response content line by line.
    ///
    ///
    /// Args:
    /// method: HTTP method
    /// url: Request URL
    /// encoding: Text encoding
    /// **kwargs: Additional request arguments
    ///
    /// Yields:
    /// Content lines as strings
    pub async fn stream_lines(&self, method: String, url: String, encoding: Option<String>, kwargs: HashMap<String, String>) -> String
    {
        // TODO: Implement
        todo!()
    }

    /// Download file with streaming and progress tracking.
    ///
    ///
    /// Args:
    /// url: URL to download from
    /// file_path: Local file path to save to
    /// chunk_size: Size of each chunk
    /// progress_callback: Optional callback for progress updates
    /// **kwargs: Additional request arguments
    pub async fn download_file(&self, url: String, file_path: String, chunk_size: Option<i64>, progress_callback: Option<String>, kwargs: HashMap<String, String>) -> ()
    {
        // TODO: Implement
        todo!()
    }

    /// Determine if a request should be retried.
    pub fn _should_retry(&self, response: Option<String>, exception: Option<Box<dyn std::error::Error + Send + Sync>>) -> bool
    {
        // TODO: This function uses external dependencies.
        // The original Python implementation should be maintained/adapted.
        //
        // Python → Rust crate mappings:
        //   httpx → reqwest
        //
        // Add these crates to Cargo.toml if implementing:
        //   reqwest = "*"
        todo!()
    }

    // Sync methods for compatibility
    /// Sync GET request.
    pub fn sync_get(&self, url: String, kwargs: HashMap<String, String>) -> String
    {
        // TODO: This function uses external dependencies.
        // The original Python implementation should be maintained/adapted.
        //
        // Python → Rust crate mappings:
        //   httpx → reqwest
        //
        // Add these crates to Cargo.toml if implementing:
        //   reqwest = "*"
        todo!()
    }

    /// Sync POST request.
    pub fn sync_post(&self, url: String, kwargs: HashMap<String, String>) -> String
    {
        // TODO: This function uses external dependencies.
        // The original Python implementation should be maintained/adapted.
        //
        // Python → Rust crate mappings:
        //   httpx → reqwest
        //
        // Add these crates to Cargo.toml if implementing:
        //   reqwest = "*"
        todo!()
    }

    /// Sync PUT request.
    pub fn sync_put(&self, url: String, kwargs: HashMap<String, String>) -> String
    {
        // TODO: This function uses external dependencies.
        // The original Python implementation should be maintained/adapted.
        //
        // Python → Rust crate mappings:
        //   httpx → reqwest
        //
        // Add these crates to Cargo.toml if implementing:
        //   reqwest = "*"
        todo!()
    }

    /// Sync PATCH request.
    pub fn sync_patch(&self, url: String, kwargs: HashMap<String, String>) -> String
    {
        // TODO: This function uses external dependencies.
        // The original Python implementation should be maintained/adapted.
        //
        // Python → Rust crate mappings:
        //   httpx → reqwest
        //
        // Add these crates to Cargo.toml if implementing:
        //   reqwest = "*"
        todo!()
    }

    /// Sync DELETE request.
    pub fn sync_delete(&self, url: String, kwargs: HashMap<String, String>) -> String
    {
        // TODO: This function uses external dependencies.
        // The original Python implementation should be maintained/adapted.
        //
        // Python → Rust crate mappings:
        //   httpx → reqwest
        //
        // Add these crates to Cargo.toml if implementing:
        //   reqwest = "*"
        todo!()
    }

    /// Sync HTTP request.
    pub fn sync_request(&self, method: String, url: String, kwargs: HashMap<String, String>) -> String
    {
        // TODO: This function uses external dependencies.
        // The original Python implementation should be maintained/adapted.
        //
        // Python → Rust crate mappings:
        //   httpx → reqwest
        //
        // Add these crates to Cargo.toml if implementing:
        //   reqwest = "*"
        todo!()
    }

}

File: src/exonware/xwsystem/http_client/advanced_client.py
#exonware/xwsystem/src/exonware/xwsystem/http_client/advanced_client.py
"""
Company: eXonware.com
Author: eXonware Backend Team
Email: connect@exonware.com
Version: 0.9.0.26
Generation Date: September 04, 2025
Advanced HTTP client with HTTP/2, streaming, pluggable transports, and modern features.
"""

from __future__ import annotations
import asyncio
import os
import ssl
from contextlib import asynccontextmanager
from dataclasses import dataclass, field
from typing import Any

from collections.abc import AsyncIterator
from urllib.parse import urljoin
from .contracts import ITransport as Transport
# Prevent httpx from importing rich (Python 3.8+ only, no legacy deps)
os.environ.setdefault("HTTPX_NO_RICH", "1")
# Import httpx - lazy installation system will handle it if missing
import httpx
from ..config.logging_setup import get_logger
from ..monitoring.error_recovery import retry_with_backoff
from .client import HttpError, RetryConfig
logger = get_logger("xwsystem.http_client.advanced_client")
@dataclass


class StreamingConfig:
    """Configuration for streaming operations."""
    chunk_size: int = 8192
    buffer_size: int = 65536
    timeout_per_chunk: float = 30.0
    max_content_length: int | None = None  # None = unlimited
@dataclass


class Http2Config:
    """Configuration for HTTP/2 features."""
    enabled: bool = True
    max_concurrent_streams: int = 100
    initial_window_size: int = 65536
    max_frame_size: int = 16384
    enable_push: bool = False
@dataclass


class AdvancedHttpConfig:
    """Advanced HTTP client configuration."""
    http2: Http2Config = field(default_factory=Http2Config)
    streaming: StreamingConfig = field(default_factory=StreamingConfig)
    retry: RetryConfig = field(default_factory=RetryConfig)
    verify_ssl: bool = True
    ssl_context: ssl.SSLContext | None = None
    trust_env: bool = True
    follow_redirects: bool = True
    max_redirects: int = 20


class MockTransport:
    """Mock transport for testing purposes."""

    def __init__(self, responses: dict[str, Any]):
        """
        Initialize mock transport.
        Args:
            responses: Dictionary mapping URLs to mock responses
        """
        self.responses = responses
        self.requests = []

    async def handle_async_request(self, request: Any) -> Any:
        """Handle async request with mock response."""
        self.requests.append(request)
        url = str(request.url)
        if url in self.responses:
            response_data = self.responses[url]
            # Create mock response
            return MockResponse(
                status_code=response_data.get('status_code', 200),
                content=response_data.get('content', b''),
                headers=response_data.get('headers', {}),
                url=url
            )
        # Default 404 response
        return MockResponse(status_code=404, content=b'Not Found', url=url)

    def handle_request(self, request: Any) -> Any:
        """Handle sync request with mock response."""
        # For simplicity, delegate to async version
        return asyncio.run(self.handle_async_request(request))


class MockResponse:
    """Mock HTTP response for testing."""

    def __init__(self, status_code: int, content: bytes, headers: dict[str, str] = None, url: str = ""):
        self.status_code = status_code
        self.content = content
        self.headers = headers or {}
        self.url = url
        self.text = content.decode('utf-8', errors='ignore')

    def json(self):
        """Parse JSON response (xwsystem JsonSerializer, no stdlib json)."""
        from exonware.xwsystem import get_serializer, JsonSerializer
        return get_serializer(JsonSerializer).loads(self.text)

    def raise_for_status(self):
        """Raise exception for HTTP error status."""
        if 400 <= self.status_code < 600:
            raise HttpError(f"HTTP {self.status_code}", self.status_code, self.text)

    async def aiter_bytes(self, chunk_size: int = 8192):
        """Async byte iterator compatible with httpx.Response."""
        for i in range(0, len(self.content), chunk_size):
            yield self.content[i:i + chunk_size]

    async def aiter_lines(self):
        """Async line iterator compatible with httpx.Response."""
        for line in self.text.splitlines():
            yield line


class AdvancedHttpClient:
    """
    Advanced HTTP client with HTTP/2, streaming, pluggable transports, and modern features.
    Features:
    - HTTP/2 support with multiplexing
    - Streaming request/response bodies
    - Pluggable transport layer for testing
    - Advanced connection management
    - Comprehensive retry and error handling
    - SSL/TLS configuration
    - Request/response hooks
    """

    def __init__(
        self,
        base_url: str | None = None,
        timeout: float = 30.0,
        config: AdvancedHttpConfig | None = None,
        transport: Transport | None = None,
        default_headers: dict[str, str] | None = None,
    ) -> None:
        """
        Initialize advanced HTTP client.
        Args:
            base_url: Base URL for all requests
            timeout: Request timeout in seconds
            config: Advanced configuration options
            transport: Custom transport (for testing/mocking)
            default_headers: Default headers for all requests
        """
        # Lazy installation system will handle httpx if missing
        self.base_url = base_url
        self.config = config or AdvancedHttpConfig()
        self.transport = transport
        self.default_headers = default_headers or {}
        # Configure HTTP/2 if enabled
        http2 = self.config.http2.enabled and not transport  # Disable HTTP/2 with custom transports
        # Configure limits
        limits = httpx.Limits(
            max_connections=100,
            max_keepalive_connections=20
        )
        # Configure SSL context
        ssl_context = self.config.ssl_context
        if ssl_context is None and self.config.verify_ssl:
            ssl_context = ssl.create_default_context()
        # Initialize clients
        client_kwargs = {
            'base_url': base_url,
            'timeout': timeout,
            'limits': limits,
            'headers': self.default_headers,
            'http2': http2,
            'verify': ssl_context if ssl_context else self.config.verify_ssl,
            'trust_env': self.config.trust_env,
            'follow_redirects': self.config.follow_redirects,
            'max_redirects': self.config.max_redirects,
        }
        if transport:
            # Use custom transport
            self._client = None
            self._async_client = None
        else:
            self._client = httpx.Client(**client_kwargs)
            self._async_client = httpx.AsyncClient(**client_kwargs)

    async def __aenter__(self) -> AdvancedHttpClient:
        """Async context manager entry."""
        return self

    async def __aexit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
        """Async context manager exit."""
        await self.aclose()

    def __enter__(self) -> AdvancedHttpClient:
        """Sync context manager entry."""
        return self

    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
        """Sync context manager exit."""
        self.close()

    def close(self) -> None:
        """Close sync client."""
        if self._client:
            self._client.close()

    async def aclose(self) -> None:
        """Close async client."""
        if self._async_client:
            await self._async_client.aclose()

    async def get(
        self,
        url: str,
        params: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
        **kwargs
    ) -> httpx.Response:
        """Async GET request."""
        return await self.request("GET", url, params=params, headers=headers, **kwargs)

    async def post(
        self,
        url: str,
        json: Any | None = None,
        data: Any | None = None,
        files: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
        **kwargs
    ) -> httpx.Response:
        """Async POST request."""
        return await self.request("POST", url, json=json, data=data, files=files, headers=headers, **kwargs)

    async def put(
        self,
        url: str,
        json: Any | None = None,
        data: Any | None = None,
        headers: dict[str, str] | None = None,
        **kwargs
    ) -> httpx.Response:
        """Async PUT request."""
        return await self.request("PUT", url, json=json, data=data, headers=headers, **kwargs)

    async def patch(
        self,
        url: str,
        json: Any | None = None,
        data: Any | None = None,
        headers: dict[str, str] | None = None,
        **kwargs
    ) -> httpx.Response:
        """Async PATCH request."""
        return await self.request("PATCH", url, json=json, data=data, headers=headers, **kwargs)

    async def delete(
        self,
        url: str,
        headers: dict[str, str] | None = None,
        **kwargs
    ) -> httpx.Response:
        """Async DELETE request."""
        return await self.request("DELETE", url, headers=headers, **kwargs)

    async def head(
        self,
        url: str,
        headers: dict[str, str] | None = None,
        **kwargs
    ) -> httpx.Response:
        """Async HEAD request."""
        return await self.request("HEAD", url, headers=headers, **kwargs)

    async def options(
        self,
        url: str,
        headers: dict[str, str] | None = None,
        **kwargs
    ) -> httpx.Response:
        """Async OPTIONS request."""
        return await self.request("OPTIONS", url, headers=headers, **kwargs)

    async def request(
        self,
        method: str,
        url: str,
        headers: dict[str, str] | None = None,
        params: dict[str, Any] | None = None,
        json: Any | None = None,
        data: Any | None = None,
        files: dict[str, Any] | None = None,
        **kwargs
    ) -> httpx.Response:
        """Make async HTTP request with retry logic."""
        # Combine headers
        request_headers = {**self.default_headers}
        if headers:
            request_headers.update(headers)
        # Prepare request arguments
        request_kwargs = {
            'method': method,
            'url': url,
            'headers': request_headers,
            'params': params,
            'files': files,
            **kwargs
        }
        if json is not None:
            request_kwargs['json'] = json
        elif data is not None:
            request_kwargs['data'] = data
        # Handle custom transport
        if self.transport:
            @retry_with_backoff(
                max_retries=self.config.retry.max_retries,
                base_delay=self.config.retry.base_delay,
                max_delay=self.config.retry.max_delay,
                backoff_factor=self.config.retry.exponential_base,
                exceptions=tuple([*self.config.retry.retry_on_exceptions, HttpError]),
            )
            async def _transport_request():
                request_url = url
                if self.base_url and isinstance(url, str) and url.startswith("/"):
                    request_url = urljoin(self.base_url, url)
                request_kwargs_for_transport = dict(request_kwargs)
                request_kwargs_for_transport["url"] = request_url
                request = httpx.Request(**request_kwargs_for_transport)
                response = await self.transport.handle_async_request(request)
                if self._should_retry(response, None):
                    raise HttpError(
                        f"HTTP {response.status_code}",
                        status_code=response.status_code,
                        response_data=getattr(response, "text", "")
                    )
                if hasattr(response, "raise_for_status"):
                    response.raise_for_status()
                return response
            return await _transport_request()
        @retry_with_backoff(
            max_retries=self.config.retry.max_retries,
            base_delay=self.config.retry.base_delay,
            max_delay=self.config.retry.max_delay,
            backoff_factor=self.config.retry.exponential_base,
            exceptions=tuple([*self.config.retry.retry_on_exceptions, HttpError]),
        )
        async def _request() -> httpx.Response:
            try:
                response = await self._async_client.request(**request_kwargs)
                if self._should_retry(response, None):
                    raise HttpError(
                        f"HTTP {response.status_code}: {response.text}",
                        status_code=response.status_code,
                        response_data=response.text
                    )
                return response
            except Exception as e:
                if self._should_retry(None, e):
                    logger.warning(f"Retrying request due to exception: {e}")
                    raise
                else:
                    raise HttpError(f"Request failed: {e}") from e
        return await _request()
    @asynccontextmanager

    async def stream(
        self,
        method: str,
        url: str,
        headers: dict[str, str] | None = None,
        params: dict[str, Any] | None = None,
        json: Any | None = None,
        data: Any | None = None,
        **kwargs
    ):
        """
        Stream HTTP request/response.
        Args:
            method: HTTP method
            url: Request URL
            headers: Request headers
            params: Query parameters
            json: JSON data
            data: Request body data
            **kwargs: Additional request arguments
        Yields:
            Streaming response object
        """
        # Combine headers
        request_headers = {**self.default_headers}
        if headers:
            request_headers.update(headers)
        # Prepare request arguments
        request_kwargs = {
            'method': method,
            'url': url,
            'headers': request_headers,
            'params': params,
            **kwargs
        }
        if json is not None:
            request_kwargs['json'] = json
        elif data is not None:
            request_kwargs['data'] = data
        if self.transport:
            # Custom transport doesn't support streaming
            response = await self.request(method, url, headers, params, json, data, **kwargs)
            yield response
            return
        async with self._async_client.stream(**request_kwargs) as response:
            yield response

    async def stream_content(
        self,
        method: str,
        url: str,
        chunk_size: int | None = None,
        **kwargs
    ) -> AsyncIterator[bytes]:
        """
        Stream response content in chunks.
        Args:
            method: HTTP method
            url: Request URL
            chunk_size: Size of each chunk
            **kwargs: Additional request arguments
        Yields:
            Content chunks as bytes
        """
        chunk_size = chunk_size or self.config.streaming.chunk_size
        async with self.stream(method, url, **kwargs) as response:
            response.raise_for_status()
            total_size = 0
            max_size = self.config.streaming.max_content_length
            async for chunk in response.aiter_bytes(chunk_size):
                if max_size and total_size + len(chunk) > max_size:
                    raise HttpError(f"Response too large: exceeds {max_size} bytes")
                total_size += len(chunk)
                yield chunk

    async def stream_lines(
        self,
        method: str,
        url: str,
        encoding: str = 'utf-8',
        **kwargs
    ) -> AsyncIterator[str]:
        """
        Stream response content line by line.
        Args:
            method: HTTP method
            url: Request URL
            encoding: Text encoding
            **kwargs: Additional request arguments
        Yields:
            Content lines as strings
        """
        async with self.stream(method, url, **kwargs) as response:
            response.raise_for_status()
            async for line in response.aiter_lines():
                yield line

    async def download_file(
        self,
        url: str,
        file_path: str,
        chunk_size: int | None = None,
        progress_callback: callable | None = None,
        **kwargs
    ) -> None:
        """
        Download file with streaming and progress tracking.
        Args:
            url: URL to download from
            file_path: Local file path to save to
            chunk_size: Size of each chunk
            progress_callback: Optional callback for progress updates
            **kwargs: Additional request arguments
        """
        from ..io.stream.async_operations import async_safe_write_bytes
        import platform
        import aiofiles
        chunk_size = chunk_size or self.config.streaming.chunk_size
        async with self.stream("GET", url, **kwargs) as response:
            response.raise_for_status()
            total_size = int(response.headers.get('content-length', 0))
            downloaded = 0
            content = b''
            async for chunk in response.aiter_bytes(chunk_size):
                content += chunk
                downloaded += len(chunk)
                if progress_callback:
                    progress_callback(downloaded, total_size)
            # Windows keeps NamedTemporaryFile handles locked; direct write avoids
            # atomic rename/remove failures when caller holds an open handle.
            if platform.system() == "Windows":
                async with aiofiles.open(file_path, "wb") as f:
                    await f.write(content)
            else:
                await async_safe_write_bytes(file_path, content)

    def _should_retry(self, response: httpx.Response | None, exception: Exception | None) -> bool:
        """Determine if a request should be retried."""
        if exception:
            return any(isinstance(exception, exc_type) for exc_type in self.config.retry.retry_on_exceptions)
        if response:
            return response.status_code in self.config.retry.retry_on_status
        return False
    # Sync wrappers (delegate to async)

    def _run_sync(self, coro):
        """Run coroutine safely from sync context (even inside an active loop)."""
        try:
            asyncio.get_running_loop()
        except RuntimeError:
            return asyncio.run(coro)

        # Already inside an event loop: run in a dedicated thread.
        import threading
        result: dict[str, Any] = {}
        error: dict[str, BaseException] = {}

        def _runner():
            try:
                result["value"] = asyncio.run(coro)
            except BaseException as exc:  # pragma: no cover
                error["exc"] = exc

        thread = threading.Thread(target=_runner, daemon=True)
        thread.start()
        thread.join()
        if "exc" in error:
            raise error["exc"]
        return result.get("value")

    def sync_get(self, url: str, **kwargs) -> httpx.Response:
        """Sync GET request."""
        return self._run_sync(self.get(url, **kwargs))

    def sync_post(self, url: str, **kwargs) -> httpx.Response:
        """Sync POST request."""
        return self._run_sync(self.post(url, **kwargs))

    def sync_put(self, url: str, **kwargs) -> httpx.Response:
        """Sync PUT request."""
        return self._run_sync(self.put(url, **kwargs))

    def sync_patch(self, url: str, **kwargs) -> httpx.Response:
        """Sync PATCH request."""
        return self._run_sync(self.patch(url, **kwargs))

    def sync_delete(self, url: str, **kwargs) -> httpx.Response:
        """Sync DELETE request."""
        return self._run_sync(self.delete(url, **kwargs))

    def sync_request(self, method: str, url: str, **kwargs) -> httpx.Response:
        """Sync HTTP request."""
        return self._run_sync(self.request(method, url, **kwargs))
